src/Feldspar/Compiler/Imperative/Frontend.hs

```diff
@@ -50,62 +50,70 @@ type Location = Maybe (Expression ())
 
 -- | Copies expressions into a destination. If the destination is
 -- an array the arguments are appended into the destination.
-copyProg :: Location -> [Expression ()] -> Program ()
-copyProg _ []      = error "copyProg: missing source parameter."
-copyProg Nothing _ = Empty
-copyProg (Just outExp) inExp =
+copyProg :: Options -> Location -> [Expression ()] -> [Program ()]
+copyProg opts _ []      = error "copyProg: missing source parameter."
+copyProg opts Nothing _ = []
+copyProg opts (Just outExp) inExp =
   case inExp of
-    [x] | outExp == x -> Empty
-    _                 -> call "copy" (map ValueParameter (outExp:inExp))
+    [x] | outExp == x -> []
+    _                 -> deepCopy opts outExp inExp
 
 -- | Expand copying of aggregate data to copying of components
-deepCopy :: Options -> [ActualParameter ()] -> [Program ()]
-deepCopy opts [ValueParameter arg1, ValueParameter arg2]
+deepCopy :: Options -> Expression () -> [Expression ()] -> [Program ()]
+deepCopy opts arg1 [arg2]
   | arg1 == arg2
   = []
 
-  | ConstExpr ArrayConst{..} <- arg2
-  = initArray (Just arg1) (litI32 $ toInteger $ length arrayValues)
-    : zipWith (\i c -> Assign (ArrayElem arg1 [litI32 i]) (ConstExpr c)) [0..] arrayValues
-
-  | NativeArray{} <- typeof arg2
-  , l@(ConstExpr (IntConst n _)) <- arrayLength arg2
-  = if n < safetyLimit opts
-      then initArray (Just arg1) l:map (\i -> Assign (ArrayElem arg1 [litI32 i]) (ArrayElem arg2 [litI32 i])) [0..(n-1)]
-      else error $ unlines ["Internal compiler error: array size (" ++ show n ++ ") too large for deepcopy", show arg1, show arg2]
-
   | StructType _ fts <- typeof arg2
   = concatMap (deepCopyField . fst) fts
 
-  | not (isArray (typeof arg1))
+  | not (isArray (typeof arg1) || isNativeArray (typeof arg1))
```

Comment:
Missing not before isNativeArray?

Reply:
No, I think the 'not' is where it should be, outside the parenthesis.

Reply:
Yeah, I'm blind.